hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java

```diff
@@ -0,0 +1,140 @@
+import com.hazelcast.jet.impl.connector.ReadMapOrCacheP.Reader;
+import com.hazelcast.partition.Partition;
+import com.hazelcast.security.permission.MapPermission;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.security.Permission;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static com.hazelcast.security.permission.ActionConstants.ACTION_CREATE;
+import static com.hazelcast.security.permission.ActionConstants.ACTION_READ;
+
+/**
+ * A meta-supplier that will compute required partitions to scan,
+ * if attribute partitioning strategy was applied.
+ */
+public abstract class SpecificPartitionsImapReaderPms<F extends CompletableFuture, B, R>
+        extends LocalProcessorMetaSupplier<F, B, R> {
+    transient int[] partitionsToScan;
+    private final List<List<Expression<?>>> requiredPartitionsExprs;
+    private transient Map<Address, int[]> partitionAssignment;
+
+    private SpecificPartitionsImapReaderPms(
+            final BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier,
+            @Nullable final List<List<Expression<?>>> requiredPartitionsExprs) {
+        super(readerSupplier);
+        this.requiredPartitionsExprs = requiredPartitionsExprs;
+    }
+
+    @Override
+    public void init(@Nonnull Context context) throws Exception {
+        super.init(context);
+        if (requiredPartitionsExprs != null) {
+            List<Integer> partitionsToScanList = new ArrayList<>();
+            partitionsToScan = new int[requiredPartitionsExprs.size()];
+            int j = 0;
+
+            HazelcastInstance hazelcastInstance = context.hazelcastInstance();
+            ExpressionEvalContext eec = ExpressionEvalContext.from(context);
+            for (List<Expression<?>> requiredPartitionsExpr : requiredPartitionsExprs) {
+                Object[] partitionKeyComponents = new Object[requiredPartitionsExpr.size()];
+                int i = 0;
+                for (Expression<?> expression : requiredPartitionsExpr) {
+                    partitionKeyComponents[i++] = expression.evalTop(null, eec);
+                }
+
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
+                );
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    return;
```

Comment:
now in this case you fail to save `partitionAssignment` -> likely NPE in `get`

Reply:
No, it will choose code path when `partitionsToScan` is null (actually, it's a default querying)

Reply:
true, but a bit unexpectected. LocalProcessorSupplier.init will obtain partitionAssignment from the context on its own.